### PR TITLE
PP-864 Allow the ability to search charges by email

### PR DESCRIPTION
**Pull request by PauloPortugal**
## WHAT

_A brief description of the pull request:_
- Allow the ability to search charges by email
  - pay-connector is already returning the `email` field, so we just need to search charges by email.
  - The email will do a partial and full match.
- Amending the unit tests for charges search by email
## HOW

_Steps to test or reproduce:_

```
cd $WORKSPACE
msl reset && msl -a up && ./run-endtoend.sh
```

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/dao/ChargeDao.java [hunk 1]
```diff
@@ -131,6 +132,8 @@ public ChargeEntity mergeAndNotifyStatusHasChanged(ChargeEntity chargeEntity) {
             predicates.add(cb.equal(charge.get(GATEWAY_ACCOUNT).get("id"), params.getGatewayAccountId()));
         if (params.getReference() != null)
             predicates.add(cb.like(charge.get(REFERENCE), '%' + params.getReference() + '%'));
+        if (params.getEmail() != null)
+            predicates.add(cb.like(charge.get(EMAIL), '%' + params.getEmail() + '%'));
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: Question: Is this really what we want to do?, does it make sense searching for emails like...? this looks like searching by id like... ❓ :neckbeard:

File: src/test/java/uk/gov/pay/connector/it/dao/ChargeDaoITest.java [hunk 2]
```diff
@@ -77,6 +77,52 @@ public void searchChargesByGatewayAccountIdOnly() throws Exception {
...
+        assertThat(charge.getId(), is(defaultTestCharge.getChargeId()));
+        assertThat(charge.getAmount(), is(defaultTestCharge.getAmount()));
+        assertThat(charge.getReference(), is(defaultTestCharge.getReference()));
+        assertThat(charge.getEmail(), is(defaultTestCharge.getEmail()));
+        assertThat(charge.getDescription(), is(DESCRIPTION));
+        assertThat(charge.getStatus(), is(defaultTestCharge.getChargeStatus().toString()));
+
+        assertDateMatch(charge.getCreatedDate().toString());
+    }
+
```
Comment (rauligs) [inline src/test/java/uk/gov/pay/connector/it/dao/ChargeDaoITest.java]: 👍

File: src/main/java/uk/gov/pay/connector/dao/ChargeDao.java (hunk 1 above)
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: Just checked the story PP-864 and there are no mention about partial/full email searches. Can you confirm it?

Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: @rauligs that is a good point, but we need to do a _partial_ and _full_ email match. 
This was according to Rory and Stephen's kick off discussion we had yesterday.

Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: @rauligs  I have sent a slack message to Rory in order for him to update the story acceptance criteria.

Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/dao/ChargeDao.java]: Ok! then I'm happy to merge. Anyway I'm interested to know about such an awkward business criteria...

**Comment by rauligs** [comment]
👍

**Pull request merged**